src/mlpack/core/data/save_image_impl.hpp

```diff
@@ -14,154 +15,94 @@
 
 // In case it hasn't been included yet.
 #include "save_image.hpp"
-#include "image_info.hpp"
 
 namespace mlpack {
 namespace data {
 
-/**
- * Save the given image to the given filename.
- *
- * @param filename Filename to save to.
- * @param matrix Matrix containing image to be saved.
- * @param info Information about the image (width/height/channels/etc.).
- * @param fatal Whether an exception should be thrown on save failure.
- */
+// Image saving API for multiple files.
 template<typename eT>
-bool Save(const std::string& filename,
-          arma::Mat<eT>& matrix,
-          ImageInfo& info,
-          const bool fatal)
-{
-  arma::Mat<unsigned char> tmpMatrix =
-      arma::conv_to<arma::Mat<unsigned char>>::from(matrix);
+bool Save(const std::vector<std::string>& files,
+          const arma::Mat<eT>& matrix,
+          ImageOptions& opts)
 
-  return SaveImage(filename, tmpMatrix, info, fatal);
+{
+  return SaveImage(files, matrix, opts);
 }
 
-// Image saving API for multiple files.
 template<typename eT>
-bool Save(const std::vector<std::string>& files,
-          arma::Mat<eT>& matrix,
-          ImageInfo& info,
-          const bool fatal)
+bool SaveImage(const std::vector<std::string>& files,
+               const arma::Mat<eT>& matrix,
+               ImageOptions& opts)
 {
-  if (files.size() == 0)
+  if (files.empty())
   {
-    if (fatal)
-    {
-      Log::Fatal << "Save(): vector of image files is empty; nothing to save."
-          << std::endl;
-    }
-    else
+    std::stringstream oss;
+    oss << "Save(): vector of image files is empty; nothing to save.";
+    HandleError(oss, opts);
+  }
+
+  // Check if we do have any type that is not supported.
+  for  (size_t i = 0; i < files.size() ; ++i)
+  {
+    if (!opts.saveType.count(Extension(files.at(i))))
     {
-      Log::Warn << "Save(): vector of image files is empty; nothing to save."
-          << std::endl;
+      std::stringstream oss;
+      oss << "Save(): file type " << opts.FileTypeToString()
+          << " isn't supported. Currently image saving supports: ";
+      for (const auto& x : opts.saveType)
+        oss << "  " << x;
+      oss << "." << std::endl;
+      HandleError(oss, opts);
     }
-
-    return false;
   }
 
-  arma::Mat<unsigned char> img;
-  bool status = true;
-
-  for (size_t i = 0; i < files.size() ; ++i)
+  size_t dimension = opts.Width() * opts.Height() * opts.Channels();
+  // We only need to check the rows since it is a matrix.
+  if (dimension != matrix.n_rows)
   {
-    arma::Mat<eT> colImg(matrix.colptr(i), matrix.n_rows, 1,
-        false, true);
-    status &= Save(files[i], colImg, info, fatal);
+    std::stringstream oss;
+    oss << "data::Save(): The given image dimensions do not match the "
+        << "dimensions of the matrix to be saved!";
```

Comment:
It's tedious but could you modify this to print the specified width/height/channels?  That would help a user a lot while debugging.  (For instance if a user forgot to specify width/height/channels, I think this check will always fail.)